flang/lib/Evaluate/intrinsics-library.cpp

```diff
@@ -633,10 +633,243 @@ static DynamicType BiggerType(DynamicType type) {
   return type;
 }
 
+/// Structure to register intrinsic argument checks that must be performed.
+using ArgumentVerifierFunc = bool (*)(
+    const std::vector<Expr<SomeType>> &, FoldingContext &);
+struct ArgumentVerifier {
+  using Key = std::string_view;
+  // Needed for implicit compare with keys.
+  constexpr operator Key() const { return key; }
+  Key key; // intrinsic name
+  ArgumentVerifierFunc verifier;
+};
+
+static constexpr int lastArg{-1};
```

Comment:
Instead of using magic values, consider using `std::optional<int>`, with `std::nullopt` signifying the last argument.  You can then access the argument via `.value_or(args.size()-1)`.

Reply:
I don't understand the advantage of this.  It seems like `std::nullopt` just becomes another kind of magic value.  Currently, the code accesses the last argument using `args.back()`, which seems pretty clear to me.